src/service/video-manager-impl.js

```diff
@@ -1232,6 +1257,209 @@ function isLandscape(win) {
 }
 
 
+/** @visibleForTesting */
+export const PERCENTAGE_INTERVAL = 5;
+
+/** @private */
+const PERCENTAGE_FREQUENCY_WHEN_PAUSED_MS = 500;
+
+/** @private */
+const PERCENTAGE_FREQUENCY_MIN_MS = 250;
+
+/** @private */
+const PERCENTAGE_FREQUENCY_MAX_MS = 4000;
+
+
+/**
+ * @param {number} durationSeconds
+ * @return {number}
+ */
+function calculateIdealPercentageFrequencyMs(durationSeconds) {
+  return durationSeconds * 10 * PERCENTAGE_INTERVAL;
+}
+
+
+/**
+ * @param {number} durationSeconds
+ * @return {number}
+ */
+function calculateActualPercentageFrequencyMs(durationSeconds) {
+  return clamp(
+      calculateIdealPercentageFrequencyMs(durationSeconds),
+      PERCENTAGE_FREQUENCY_MIN_MS,
+      PERCENTAGE_FREQUENCY_MAX_MS);
+}
+
+
+/** @visibleForTesting */
+export class AnalyticsPercentageTracker {
+  /**
+   * @param {!Window} win
+   * @param {!VideoEntry} entry
+   */
+  constructor(win, entry) {
+
+    // These are destructured in `calculate_()`, but the linter thinks they're
+    // unused.
+    /* eslint-disable amphtml-internal/unused-private-field */
+
+    /** @private @const {!./timer-impl.Timer} */
+    this.timer_ = Services.timerFor(win);
+
+    /** @private @const {!VideoEntry} */
+    this.entry_ = entry;
+
+    /* eslint-enable amphtml-internal/unused-private-field */
+
+    /** @private {?Array<!UnlistenDef>} */
+    this.unlisteners_ = null;
+
+    /** @private {number} */
+    this.last_ = 0;
+
+    /** @private {number} */
+    this.triggerId_ = 0;
+  }
+
+  /** @public */
+  start() {
+    const {element} = this.entry_.video;
+
+    this.stop();
+
+    this.unlisteners_ = this.unlisteners_ || [];
+
+    this.unlisteners_.push(
+        listenOnce(element, VideoEvents.LOADEDMETADATA, () => {
+          if (this.hasDuration_()) {
+            this.calculate_(this.triggerId_);
+          }
+        }),
+
+        listen(element, VideoEvents.ENDED, () => {
+          if (this.hasDuration_()) {
+            this.maybeTrigger_(/* normalizedPercentage */ 100);
+          }
+        }));
+  }
+
+  /** @public */
+  stop() {
+    if (!this.unlisteners_) {
+      return;
+    }
+    while (this.unlisteners_.length > 0) {
+      this.unlisteners_.pop().call();
```

Comment:
If you look at the signature above you'll see that unlisteners is a set of `!UnlistenDef` (effectively just 1st order fns). This empties the array while also executing all unlisteners.

Reply:
Ah that makes sense, I didn't expect `call` to be used this way. I think it might help readability to call the function conventionally instead of with `call`. It's up to your best judgment

```
const unlisten = this.unlisteners_.pop();
unlisten();
```